### PR TITLE
Add partial result broker metrics (Based on PR #868)

### PR DESCRIPTION
**Pull request by Jackie-Jiang**
1. Add metrics to track processing exception and partial servers responded.
   2. Add metrics to track entries scanned in/post filter.
   3. Add numServersQueried/Responded into broker response.
   4. Enhance server CombinePlanNode to pass out exceptions.

### CODE REVIEW

File: pinot-common/src/main/java/com/linkedin/pinot/common/metrics/BrokerMeter.java [hunk 1]
```diff
@@ -28,10 +28,31 @@
...
+  NO_SERVER_FOUND_EXCEPTIONS("exceptions", false),
+  // Gather phase.
+  RESPONSE_FETCH_EXCEPTIONS("exceptions", false),
+  // Response deserialize phase.
+  DATA_TABLE_DESERIALIZATION_EXCEPTIONS("exceptions", false),
+  // Reduce responses phase.
+  RESPONSE_MERGE_EXCEPTIONS("exceptions", false),
+
+  // These metrics track the number of partial results.
+  PROCESSING_EXCEPTION_CAUGHT("exceptions", false),
```
Comment (antumbde) [inline pinot-common/src/main/java/com/linkedin/pinot/common/metrics/BrokerMeter.java]: pls document what these metrics mean. From name, I can not figure what is exactly tracked by processing_exception_caught.

Why "caught" part in the name ?

Comment (Jackie-Jiang) [inline pinot-common/src/main/java/com/linkedin/pinot/common/metrics/BrokerMeter.java]: Updated the comments to make it more clear.

File: pinot-core/src/main/java/com/linkedin/pinot/core/plan/CombinePlanNode.java [hunk 2]
```diff
@@ -76,37 +73,39 @@ public Operator run() {
...
+        }));
       }
+
+      // Try to get results from all jobs. Cancel all remaining jobs if caught any exception.
+      int index = 0;
       try {
-        // Check if count reached zero in the time limit.
-        if (!latch.await(TIME_OUT_IN_SECONDS_FOR_PARALLEL_RUN, TimeUnit.SECONDS)) {
-          throw new RuntimeException(QueryException.COMBINE_SEGMENT_PLAN_TIMEOUT_ERROR);
+        while (index < numPlanNodes) {
```
Comment (mayankshriv) [inline pinot-core/src/main/java/com/linkedin/pinot/core/plan/CombinePlanNode.java]: Could you please add a unit test, specifically testing for server side timeout?

Comment (Jackie-Jiang) [inline pinot-core/src/main/java/com/linkedin/pinot/core/plan/CombinePlanNode.java]: Added.

**Pull request merged**